### PR TITLE
Docs: Rearrange AMPI manual

### PR DESCRIPTION
**Pull request by evan-charmworks**
May be best viewed in two ways:

1. By examining the diff, using `git diff -w --color-moved --no-merges master...`.
2. By using a tool such as _restview_ to preview the result in a web browser. (Most useful for comparing the table of contents.)

### CODE REVIEW

File: doc/ampi/manual.rst [hunk 1]
```diff
@@ -213,79 +213,198 @@ Non-blocking collectives have been defined in AMPI since before
...
+The build script for Charm++ is called ``build``. The syntax for this
+script is:
 
-.. code-block:: fortran
+.. code-block:: bash
 
-   ! Fortran:
-   integer :: my_wth, flag, ierr
-   call MPI_Comm_get_attr(MPI_COMM_WORLD, AMPI_MY_WTH, my_wth, flag, ierr)
+   $ build <target> <version> <opts>
```
Comment (matthiasdiener) [inline doc/ampi/manual.rst]: should be ./build

File: doc/ampi/manual.rst [hunk 2]
```diff
@@ -213,79 +213,198 @@ Non-blocking collectives have been defined in AMPI since before
...
+``AMPI_SMP_RDMA_THRESHOLD_DEFAULT``. Contiguous messages with sizes
+larger than the threshold are sent via RDMA on communication layers that
+support this capability. You can also set the environment variables
+``AMPI_RDMA_THRESHOLD`` and ``AMPI_SMP_RDMA_THRESHOLD`` before running a
+job to override the default specified at build time.
+
+Building AMPI Programs
+----------------------
+
+AMPI provides compiler wrappers such as *ampicc*, *ampif90*, and
```
Comment (matthiasdiener) [inline doc/ampi/manual.rst]: maybe: `ampicc`   etc

File: doc/ampi/manual.rst [hunk 3]
```diff
@@ -213,79 +213,198 @@ Non-blocking collectives have been defined in AMPI since before
...
 
-   call AMPI_Command_argument_count(argc)
-   allocate(raw_arguments(argc))
-   do i = 1, size(raw_arguments)
-       call AMPI_Get_command_argument(i, raw_arguments(i), arg_len, ierr)
-   end do
+   $ ./build AMPI gni-crayxc --with-production --disable-ampi-error-checking
 
-Name for Main Program
+AMPI can also be built with support for shared memory on any
```
Comment (matthiasdiener) [inline doc/ampi/manual.rst]: could mention here what shared memory is used for

File: doc/ampi/manual.rst [hunk 4]
```diff
@@ -213,79 +213,198 @@ Non-blocking collectives have been defined in AMPI since before
...
+
+.. code-block:: bash
+
+   $ ./ampirun -np 16 ./pgm +balancer RefineLB
+
+Other options
+~~~~~~~~~~~~~
+
+Note that for AMPI programs compiled with gfortran, users may need to
+set the following environment variable to see program output to stdout:
```
Comment (matthiasdiener) [inline doc/ampi/manual.rst]: maybe: on stdout

File: doc/ampi/manual.rst [hunk 5]
```diff
@@ -307,9 +426,37 @@ C or C++
 
 The main function can be left as is, if ``mpi.h`` is included before the
 main function. This header file has a preprocessor macro that renames
-main, and the renamed version is called by the AMPI runtime by each
+main, and the renamed version is called by the AMPI runtime for each
 thread.
```
Comment (matthiasdiener) [inline doc/ampi/manual.rst]: maybe replace thread with rank

File: doc/ampi/manual.rst [hunk 6]
```diff
@@ -747,13 +894,38 @@ different schemes.
    FS-Globals           Yes Yes    Yes    No   Maybe   Yes   Yes
    ==================== === ====== ====== ==== ======= ===== =====
```
Comment (matthiasdiener) [inline doc/ampi/manual.rst]: The table could use some improvements:

- x86, x86_64 -> Linux x86 (32/64 bit) (same for PPC, ARM7)
- No -> Not supported
- Maybe -> Not implemented (yet)
- Transformation -> Source-to-Source Transformation
- order in the same way as they are discussed in the manual

File: doc/ampi/manual.rst [hunk 7]
```diff
@@ -766,119 +938,29 @@ For systems that do not support Isomalloc and for users that wish to
 have more fine-grain control over which application data structures will
 be copied at migration time, we have added a few calls to AMPI. These
 include the ability to register thread-specific data with the run-time
-system, to pack and unpack all of the thread’s data, and to express
-willingness to migrate.
+system, and the means to pack and unpack all of the thread’s data.
 
-Registering User Data
-~~~~~~~~~~~~~~~~~~~~~
+**Most users may skip this section unless you have specific needs.**
```
Comment (matthiasdiener) [inline doc/ampi/manual.rst]: maybe use `.. warning::`

File: doc/ampi/manual.rst (hunk 2 above)
Comment (evan-charmworks) [reply doc/ampi/manual.rst]: I don't think "`ampicc` etc" will be clear enough for first-time readers.

File: doc/ampi/manual.rst (hunk 1 above)
Comment (evan-charmworks) [reply doc/ampi/manual.rst]: Done

File: doc/ampi/manual.rst (hunk 3 above)
Comment (evan-charmworks) [reply doc/ampi/manual.rst]: Done

File: doc/ampi/manual.rst (hunk 4 above)
Comment (evan-charmworks) [reply doc/ampi/manual.rst]: Done

File: doc/ampi/manual.rst (hunk 5 above)
Comment (evan-charmworks) [reply doc/ampi/manual.rst]: Done

File: doc/ampi/manual.rst (hunk 6 above)
Comment (evan-charmworks) [reply doc/ampi/manual.rst]: I added a Linux column and moved the x86 architectures to be next to the others. Specifying OS and arch in the same column would be too verbose.

"No" and "Maybe" are already explained in the table's key, and changing them would blow up the size.

Changed "Transformation" to "Manual Change" and reordered the rows to match the document's text.

File: doc/ampi/manual.rst (hunk 7 above)
Comment (evan-charmworks) [reply doc/ampi/manual.rst]: Done

File: doc/ampi/manual.rst (hunk 2 above)
Comment (evan-charmworks) [reply doc/ampi/manual.rst]: I don't think "`ampicc` etc" will be clear enough for first-time readers.

Comment (matthiasdiener) [reply doc/ampi/manual.rst]: Sorry, what I meant here was, replacing `*ampicc* ` with ` ``ampicc`` `, `*ampif90* ` with ` ``ampif90`` `, and so on.

Comment (evan-charmworks) [reply doc/ampi/manual.rst]: Done

File: doc/ampi/manual.rst (hunk 6 above)
Comment (matthiasdiener) [reply doc/ampi/manual.rst]: > I added a Linux column and moved the x86 architectures to be next to the others. Specifying OS and arch in the same column would be too verbose.
> 
> "No" and "Maybe" are already explained in the table's key, and changing them would blow up the size.

+1.

> Changed "Transformation" to "Manual Change" and reordered the rows to match the document's text.

My concern is that the connection to the manual section where this transformation is explained isn't too clear. Maybe change the table entry and subsection title to "Source code change"?

Comment (evan-charmworks) [reply doc/ampi/manual.rst]: Changed the Manual Change table entry and section title to Manual Code Editing.

**Comment by evan-charmworks** [comment]
Merging this will provide potential new AMPI users with a less confusing and outdated primary source of documentation.

**Pull request merged**